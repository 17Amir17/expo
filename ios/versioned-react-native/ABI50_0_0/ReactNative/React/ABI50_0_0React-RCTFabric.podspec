# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

require "json"

package = JSON.parse(File.read(File.join(__dir__, "..", "package.json")))
version = package['version']



folly_flags = '-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -DFOLLY_CFG_NO_COROUTINES=1'
folly_compiler_flags = folly_flags + ' ' + '-Wno-comma -Wno-shorten-64-to-32'
folly_version = '2022.05.16.00'
boost_compiler_flags = '-Wno-documentation'

header_search_paths = [
  "\"$(PODS_TARGET_SRCROOT)/ReactCommon\"",
  "\"$(PODS_ROOT)/boost\"",
  "\"$(PODS_ROOT)/DoubleConversion\"",
  "\"$(PODS_ROOT)/fmt/include\"",
  "\"$(PODS_ROOT)/RCT-Folly\"",
  "\"$(PODS_ROOT)/Headers/Private/ABI50_0_0React-Core\"",
  "\"$(PODS_ROOT)/Headers/Private/ABI50_0_0Yoga\"",
  "\"$(PODS_ROOT)/Headers/Public/React-Codegen\"",
]

if ENV['USE_FRAMEWORKS']
  create_header_search_path_for_frameworks("React-RCTFabric", :framework_name => "RCTFabric")
    .each { |search_path| header_search_paths << "\"#{search_path}\""}
end

module_name = "ABI50_0_0RCTFabric"
header_dir = "ABI50_0_0React"

Pod::Spec.new do |s|
  s.name                   = "ABI50_0_0React-RCTFabric"
  s.version                = version
  s.summary                = "RCTFabric for React Native."
  s.homepage               = "https://reactnative.dev/"
  s.license                = package["license"]
  s.author                 = "Meta Platforms, Inc. and its affiliates"
  s.platforms              = min_supported_versions
  s.source                 = { :path => "." }
  s.source_files           = "Fabric/**/*.{c,h,m,mm,S,cpp}"
  s.exclude_files          = "**/tests/*",
                             "**/android/*",
  s.compiler_flags         = folly_compiler_flags + ' ' + boost_compiler_flags
  s.header_dir             = header_dir
  s.module_name            = module_name
  s.framework              = ["JavaScriptCore", "MobileCoreServices"]
  s.pod_target_xcconfig    = {
    "HEADER_SEARCH_PATHS" => header_search_paths,
    "OTHER_CFLAGS" => "$(inherited) -DRN_FABRIC_ENABLED" + " " + folly_flags,
    "CLANG_CXX_LANGUAGE_STANDARD" => "c++20"
  }.merge!(ENV['USE_FRAMEWORKS'] != nil ? {
    "PUBLIC_HEADERS_FOLDER_PATH" => "#{module_name}.framework/Headers/#{header_dir}"
  }: {})

  s.dependency "ABI50_0_0React-Core"
  s.dependency "ABI50_0_0React-RCTImage"
  s.dependency "RCT-Folly/Fabric", folly_version
  s.dependency "glog"
  s.dependency "ABI50_0_0Yoga"
  s.dependency "ABI50_0_0React-RCTText"
  s.dependency "ABI50_0_0React-jsi"

  add_dependency(s, "ABI50_0_0React-FabricImage")
  add_dependency(s, "ABI50_0_0React-Fabric", :additional_framework_paths => [
    "react/renderer/textlayoutmanager/platform/ios",
    "react/renderer/components/textinput/iostextinput",
    "react/renderer/components/view/platform/cxx",
    "react/renderer/imagemanager/platform/ios",
  ])
  add_dependency(s, "ABI50_0_0React-nativeconfig")
  add_dependency(s, "ABI50_0_0React-graphics", :additional_framework_paths => ["react/renderer/graphics/platform/ios"])
  add_dependency(s, "ABI50_0_0React-ImageManager")
  add_dependency(s, "ABI50_0_0React-debug")
  add_dependency(s, "ABI50_0_0React-utils")
  add_dependency(s, "ABI50_0_0React-rendererdebug")
  add_dependency(s, "ABI50_0_0React-runtimescheduler")

  if ENV["USE_HERMES"] == nil || ENV["USE_HERMES"] == "1"
    s.dependency "ABI50_0_0hermes-engine"
  else
    s.dependency "ABI50_0_0React-jsc"
  end

  s.test_spec 'Tests' do |test_spec|
    test_spec.source_files = "Tests/**/*.{mm}"
    test_spec.framework = "XCTest"
  end
end
