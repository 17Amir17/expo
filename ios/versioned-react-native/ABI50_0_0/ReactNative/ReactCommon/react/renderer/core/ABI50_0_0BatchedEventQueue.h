/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#pragma once

#include <ABI50_0_0React/renderer/core/ABI50_0_0EventQueue.h>
#include <ABI50_0_0React/renderer/core/ABI50_0_0EventQueueProcessor.h>

namespace ABI50_0_0facebook::ABI50_0_0React {

/*
 * Event Queue that dispatches event in batches synchronizing them with
 * an Event Beat.
 */
class BatchedEventQueue final : public EventQueue {
 public:
  BatchedEventQueue(
      EventQueueProcessor eventProcessor,
      std::unique_ptr<EventBeat> eventBeat);

  void onEnqueue() const override;
};

} // namespace ABI50_0_0facebook::ABI50_0_0React
