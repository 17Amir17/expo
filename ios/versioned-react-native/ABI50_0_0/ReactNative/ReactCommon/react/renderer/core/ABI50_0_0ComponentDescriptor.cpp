/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "ABI50_0_0ComponentDescriptor.h"

namespace ABI50_0_0facebook::ABI50_0_0React {

ComponentDescriptor::ComponentDescriptor(
    const ComponentDescriptorParameters& parameters)
    : eventDispatcher_(parameters.eventDispatcher),
      contextContainer_(parameters.contextContainer),
      flavor_(parameters.flavor) {}

const ContextContainer::Shared& ComponentDescriptor::getContextContainer()
    const {
  return contextContainer_;
}

} // namespace ABI50_0_0facebook::ABI50_0_0React
