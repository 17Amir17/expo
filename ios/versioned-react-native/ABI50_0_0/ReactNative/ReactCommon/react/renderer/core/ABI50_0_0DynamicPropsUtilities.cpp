/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "ABI50_0_0DynamicPropsUtilities.h"

namespace ABI50_0_0facebook::ABI50_0_0React {

folly::dynamic mergeDynamicProps(
    const folly::dynamic& source,
    const folly::dynamic& patch) {
  auto result = source;

  if (!result.isObject()) {
    result = folly::dynamic::object();
  }

  if (!patch.isObject()) {
    return result;
  }

  // Note, here we have to preserve sub-prop objects with `null` value as
  // an indication for the legacy mounting layer that it needs to clean them up.
  for (const auto& pair : patch.items()) {
    result[pair.first] = pair.second;
  }

  return result;
}

} // namespace ABI50_0_0facebook::ABI50_0_0React
