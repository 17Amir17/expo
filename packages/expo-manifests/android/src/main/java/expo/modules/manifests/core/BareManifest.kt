package expo.modules.manifests.core

import expo.modules.jsonutils.require
import org.json.JSONException
import org.json.JSONObject

data class BareManifest(private val json: JSONObject) : BaseLegacyManifest {
  override fun getRawJson(): JSONObject {
    return json
  }

  @Deprecated(message = "Prefer to use specific field getters")
  override fun toString(): String {
    return getRawJson().toString()
  }

  /**
   * A UUID for this manifest.
   */
  @Throws(JSONException::class)
  fun getID(): String = json.require("id")

  @Throws(JSONException::class)
  fun getCommitTimeLong(): Long = json.require("commitTime")
}
