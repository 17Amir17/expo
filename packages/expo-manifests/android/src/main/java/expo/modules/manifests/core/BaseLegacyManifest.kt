package expo.modules.manifests.core

import expo.modules.jsonutils.getNullable
import expo.modules.jsonutils.require
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

interface BaseLegacyManifest : Manifest {
  override fun getStableLegacyID(): String = getRawJson().getNullable("originalFullName") ?: getLegacyID()

  override fun getScopeKey(): String = getRawJson().getNullable("scopeKey") ?: getStableLegacyID()

  override fun getEASProjectID(): String? = getRawJson().getNullable("projectId")

  override fun getAssets(): JSONArray? = getRawJson().getNullable("assets")

  @Throws(JSONException::class)
  override fun getBundleURL(): String = getRawJson().require("bundleUrl")

  override fun getExpoGoSDKVersion(): String? = getRawJson().getNullable("sdkVersion")

  override fun getExpoGoConfigRootObject(): JSONObject? {
    return getRawJson()
  }

  override fun getExpoClientConfigRootObject(): JSONObject? {
    return getRawJson()
  }

  override fun getSlug(): String? = getRawJson().getNullable("slug")

  override fun getAppKey(): String? = getRawJson().getNullable("appKey")

  fun getCommitTime(): String? = getRawJson().getNullable("commitTime")
}
