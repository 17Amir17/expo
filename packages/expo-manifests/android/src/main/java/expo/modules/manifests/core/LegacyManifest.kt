package expo.modules.manifests.core

import expo.modules.jsonutils.getNullable
import expo.modules.jsonutils.require
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

data class LegacyManifest(private val json: JSONObject) : BaseLegacyManifest {
  override fun getRawJson(): JSONObject {
    return json
  }

  @Deprecated(message = "Prefer to use specific field getters")
  override fun toString(): String {
    return getRawJson().toString()
  }

  @Throws(JSONException::class)
  fun getBundleKey(): String? = json.getNullable("bundleKey")

  @Throws(JSONException::class)
  fun getReleaseId(): String = json.require("releaseId")

  fun getRuntimeVersion(): String? = json.getNullable("runtimeVersion")

  @Throws(JSONException::class)
  fun getBundledAssets(): JSONArray? = json.getNullable("bundledAssets")

  open fun getAssetUrlOverride(): String? = json.getNullable("assetUrlOverride")
}
