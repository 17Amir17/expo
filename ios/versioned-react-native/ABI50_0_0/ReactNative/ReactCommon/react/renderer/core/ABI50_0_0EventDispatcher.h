/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#pragma once

#include <ABI50_0_0React/renderer/core/ABI50_0_0BatchedEventQueue.h>
#include <ABI50_0_0React/renderer/core/ABI50_0_0EventBeat.h>
#include <ABI50_0_0React/renderer/core/ABI50_0_0EventListener.h>
#include <ABI50_0_0React/renderer/core/ABI50_0_0EventPriority.h>
#include <ABI50_0_0React/renderer/core/ABI50_0_0EventQueueProcessor.h>
#include <ABI50_0_0React/renderer/core/ABI50_0_0StateUpdate.h>
#include <ABI50_0_0React/renderer/core/ABI50_0_0UnbatchedEventQueue.h>

namespace ABI50_0_0facebook::ABI50_0_0React {

struct RawEvent;

/*
 * Represents event-delivery infrastructure.
 * Particular `EventEmitter` clases use this for sending events.
 */
class EventDispatcher {
 public:
  using Shared = std::shared_ptr<const EventDispatcher>;
  using Weak = std::weak_ptr<const EventDispatcher>;

  EventDispatcher(
      const EventQueueProcessor& eventProcessor,
      const EventBeat::Factory& synchonousEventBeatFactory,
      const EventBeat::Factory& asynchronousEventBeatFactory,
      const EventBeat::SharedOwnerBox& ownerBox);

  /*
   * Dispatches a raw event with given priority using event-delivery pipe.
   */
  void dispatchEvent(RawEvent&& rawEvent, EventPriority priority) const;

  /*
   * Dispatches a raw event with asynchronous batched priority. Before the
   * dispatch we make sure that no other RawEvent of same type and same target
   * is on the queue.
   */
  void dispatchUniqueEvent(RawEvent&& rawEvent) const;

  /*
   * Dispatches a state update with given priority.
   */
  void dispatchStateUpdate(StateUpdate&& stateUpdate, EventPriority priority)
      const;

#pragma mark - Event listeners
  /*
   * Adds provided event listener to the event dispatcher.
   */
  void addListener(const std::shared_ptr<const EventListener>& listener) const;

  /*
   * Removes provided event listener to the event dispatcher.
   */
  void removeListener(
      const std::shared_ptr<const EventListener>& listener) const;

 private:
  const EventQueue& getEventQueue(EventPriority priority) const;

  std::unique_ptr<UnbatchedEventQueue> synchronousUnbatchedQueue_;
  std::unique_ptr<BatchedEventQueue> synchronousBatchedQueue_;
  std::unique_ptr<UnbatchedEventQueue> asynchronousUnbatchedQueue_;
  std::unique_ptr<BatchedEventQueue> asynchronousBatchedQueue_;

  mutable EventListenerContainer eventListeners_;
};

} // namespace ABI50_0_0facebook::ABI50_0_0React
