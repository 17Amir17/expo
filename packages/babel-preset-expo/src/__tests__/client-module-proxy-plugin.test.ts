import * as babel from '@babel/core';

import preset from '..';

const ENABLED_CALLER = {
  name: 'metro',
  isDev: false,
  isServer: false,
  projectRoot: '/',
};

function getCaller(props: Record<string, string | boolean>): babel.TransformCaller {
  return props as unknown as babel.TransformCaller;
}

const DEF_OPTIONS = {
  // Ensure this is absolute to prevent the filename from being converted to absolute and breaking CI tests.
  filename: '/unknown',

  babelrc: false,
  presets: [[preset, { disableImportExportTransform: true }]],
  sourceMaps: true,
  configFile: false,
  compact: false,
  comments: true,
  retainLines: false,
  caller: getCaller({ ...ENABLED_CALLER, platform: 'ios' }),
};

const originalEnv = process.env;

beforeEach(() => {
  process.env = { ...originalEnv };
});

afterAll(() => {
  process.env = { ...originalEnv };
});

function getOpts(caller: Record<string, string | boolean>) {
  return {
    ...DEF_OPTIONS,
    caller: getCaller({ ...ENABLED_CALLER, ...caller }),
  };
}

describe('forbidden server APIs', () => {
  function runServerPass(src: string) {
    return babel.transform(
      src,
      getOpts({
        isReactServer: true,
        platform: 'ios',
      })
    );
  }
  it(`does not assert importing client-side APIs in client components (react server mode)`, () => {
    // This test covers the order of server registry running before the assertion to remove the import.
    expect(runServerPass(`"use client"; import { useState } from 'react';`).code).toMatch(
      'react-server-dom-webpack'
    );
  });
  it(`asserts importing client-side React APIs in server components`, () => {
    expect(() => runServerPass(`import { useState } from 'react';`)).toThrowErrorMatchingSnapshot();
    expect(() => runServerPass(`import { useRef, useContext } from 'react';`)).toThrowError();
    expect(() => runServerPass(`import { PureComponent } from 'react';`)).toThrowError();
    expect(() => runServerPass(`import { Component } from 'react';`)).toThrowError();
    expect(() => runServerPass(`import { useRandom } from 'react';`)).not.toThrowError();
  });
  it(`asserts importing client-side react-dom APIs in server components`, () => {
    expect(() =>
      runServerPass(`import { findDOMNode } from 'react-dom';`)
    ).toThrowErrorMatchingSnapshot();
    expect(() => runServerPass(`import { useRandom } from 'react-dom';`)).not.toThrowError();
  });
  it(`does not assert importing client-side react-dom APIs in server components if they are in node modules`, () => {
    expect(
      babel.transform(`import { findDOMNode } from 'react-dom';`, {
        ...DEF_OPTIONS,
        filename: '/bacon/node_modules/@bacons/breakfast.js',
        caller: getCaller({ ...ENABLED_CALLER, isReactServer: true, platform: 'ios' }),
      }).code
    ).toBe(`import { findDOMNode } from 'react-dom';`);
  });

  it(`asserts client-side React API usage in server components`, () => {
    expect(() =>
      runServerPass(`
    import * as React from 'react';
    
    export default function App() {
      const [state, setState] = React.useState(0);
      return <div>{state}</div>;
    }
    `)
    ).toThrow(/cannot be used in a React server component/);
  });
  it(`asserts client-side React API usage in server components (default import)`, () => {
    expect(() =>
      runServerPass(`
    import React from 'react';
    
    export default function App() {
      const ref = React.useRef(null);
      return <div>{ref}</div>;
    }
    `)
    ).toThrow(/cannot be used in a React server component/);
  });
  it(`asserts client-side React class component usage in server components`, () => {
    expect(() =>
      runServerPass(`
    import React from 'react';
    
    class App extends React.Component {
        render() {
        return <div />;
        }
    }
    `)
    ).toThrow(/Class components cannot be/);
  });

  it(`allows client-side React API usage in client components`, () => {
    runServerPass(`
    "use client"
    import React from 'react';
    
    export default function App() {
      const ref = React.useRef(null);
      return <div>{ref}</div>;
    }
    `);
    runServerPass(`
    "use client"
    import * as React from 'react';
    
    export default function App() {
      const ref = React.useRef(null);
      return <div>{ref}</div>;
    }
    `);
  });

  it(`allows client-side React class component usage in client components`, () => {
    runServerPass(`
      "use client"
    import React from 'react';
    
    class App extends React.Component {
        render() {
        return <div />;
        }
    }
    `);
  });
});

describe('use client', () => {
  it(`does nothing without use client directive`, () => {
    const options = {
      ...DEF_OPTIONS,
      caller: getCaller({ ...ENABLED_CALLER, isReactServer: true, platform: 'ios' }),
    };

    const sourceCode = `
      export const foo = 'bar';
    `;

    const contents = babel.transform(sourceCode, options)!.code;

    expect(contents).toMatchSnapshot();
    expect(contents).not.toMatch('react-server-dom-webpack');
  });
  it(`collects metadata with React client references`, () => {
    const options = {
      ...DEF_OPTIONS,
      caller: getCaller({ ...ENABLED_CALLER, isReactServer: false, platform: 'ios' }),
    };

    const sourceCode = `
    "use client";
    import { Text } from 'react-native';
    
    export const foo = 'bar';
    
    export default function App() {
      return <Text>Hello World</Text>
    }
    `;

    const contents = babel.transform(sourceCode, options);
    expect(contents.metadata).toEqual({
      clientReferences: { entryPoint: 'file:///unknown', exports: ['foo', 'default'] },
    });

    // This isn't added because the process is not react server.
    expect(contents.code).not.toMatch('react-server-dom-webpack');
  });

  it(`replaces client exports with React client references`, () => {
    const options = {
      ...DEF_OPTIONS,
      caller: getCaller({ ...ENABLED_CALLER, isReactServer: true, platform: 'ios' }),
    };

    const sourceCode = `
  "use client";
  import { Text } from 'react-native';
  
  export const foo = 'bar';
  
  export default function App() {
    return <Text>Hello World</Text>
  }
  `;

    const contents = babel.transform(sourceCode, options)!.code;
    expect(contents).toMatchSnapshot();

    expect(contents).toMatch('react-server-dom-webpack');
  });

  it(`asserts that use client and use server cannot be used together`, () => {
    const options = {
      ...DEF_OPTIONS,
      caller: getCaller({ ...ENABLED_CALLER, isReactServer: true, platform: 'ios' }),
    };

    const sourceCode = `
    'use server';
    'use client';
    
    export const greet = (name: string) => \`Hello $\{name} from server!\`;
    `;

    expect(() => babel.transform(sourceCode, options)).toThrowErrorMatchingSnapshot();
  });
});

describe('use server', () => {
  it(`replaces server action exports with React server references`, () => {
    const options = {
      ...DEF_OPTIONS,
      caller: getCaller({ ...ENABLED_CALLER, isReactServer: true, platform: 'ios' }),
    };

    const sourceCode = `
        'use server';
      
        export const greet = (name: string) => \`Hello $\{name} from server!\`;
      `;

    const contents = babel.transform(sourceCode, options)!.code;
    expect(contents).toMatchSnapshot();
  });
});
