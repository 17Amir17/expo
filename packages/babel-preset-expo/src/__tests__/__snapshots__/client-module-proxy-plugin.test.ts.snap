// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forbidden server APIs asserts importing client-side React APIs in server components 1`] = `
"/unknown: Client-only "react" API "useState" cannot be imported in a React server component. Add the "use client" directive to the top of this file or one of the parent files to enable running this stateful code on a user's device.
[0m[31m[1m>[22m[39m[90m 1 |[39m [36mimport[39m { useState } [36mfrom[39m [32m'react'[39m[33m;[39m[0m
[0m [90m   |[39m [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[0m"
`;

exports[`forbidden server APIs asserts importing client-side react-dom APIs in server components 1`] = `
"/unknown: Client-only "react-dom" API "findDOMNode" cannot be imported in a React server component. Add the "use client" directive to the top of this file or one of the parent files to enable running this stateful code on a user's device.
[0m[31m[1m>[22m[39m[90m 1 |[39m [36mimport[39m { findDOMNode } [36mfrom[39m [32m'react-dom'[39m[33m;[39m[0m
[0m [90m   |[39m [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[0m"
`;

exports[`use client asserts that use client and use server cannot be used together 1`] = `
"/unknown: It's not possible to have both \`use client\` and \`use server\` directives in the same file.
[0m[31m[1m>[22m[39m[90m 1 |[39m[0m
[0m [90m   |[39m [31m[1m^[22m[39m[0m
[0m [90m 2 |[39m     [32m'use server'[39m[33m;[39m[0m
[0m [90m 3 |[39m     [32m'use client'[39m[33m;[39m[0m
[0m [90m 4 |[39m     [0m"
`;

exports[`use client does nothing without use client directive 1`] = `"export var foo = 'bar';"`;

exports[`use client replaces client exports with React client references 1`] = `"module.exports = require("react-server-dom-webpack/server").createClientModuleProxy("file:///unknown");"`;

exports[`use server replaces server action exports with React server references 1`] = `
"'use server';

export var greet = function greet(name) {
  return \`Hello \${name} from server!\`;
};
(function () {
  if (typeof module.exports === "function") {
    require("react-server-dom-webpack/server").registerServerReference(module.exports, "file:///unknown", null);
  } else {
    for (var key in module.exports) {
      if (typeof module.exports[key] === "function") require("react-server-dom-webpack/server").registerServerReference(module.exports[key], "file:///unknown", key);
    }
  }
})();"
`;
